crates/hir-ty/src/diagnostics/expr.rs

```diff
@@ -290,7 +290,28 @@ impl ExprValidator {
 
     fn check_for_unnecessary_else(&mut self, id: ExprId, expr: &Expr, body: &Body) {
         if let Expr::If { condition: _, then_branch, else_branch } = expr {
-            if else_branch.is_none() {
+            if let Some(else_branch) = else_branch {
+                // If else branch has a tail, it is an "expression" that produces a value,
+                // e.g. `let a = if { ... } else { ... };` and this `else` is not unnecessary
+                let mut branch = *else_branch;
+                loop {
+                    match body.exprs[branch] {
+                        Expr::Block { tail: Some(_), .. } => return,
+                        Expr::If { then_branch, else_branch, .. } => {
+                            if let Expr::Block { tail: Some(_), .. } = body.exprs[then_branch] {
+                                return;
+                            }
+                            if let Some(else_branch) = else_branch {
+                                // Continue checking for branches like `if { ... } else if { ... } else...`
+                                branch = else_branch;
+                                continue;
+                            }
+                        }
+                        _ => break,
+                    }
+                    break;
+                }
+            } else {
                 return;
             }
```

Comment:
> you should instead be checking whether the if expression has a let statement as a "direct ancestor"

Yes, this would be the right way for handling `let a = if ...` cases and I had considered that method - though quite ugly way iterating exprs once more for let bindings 😅, not as sophisticated as yours(I've learned a lot from your suggestion codes) - but I think that we should also filter out the cases that if-else expression is a tail expression of the other block, like the following example.

```rust
let foo = match bar {
    ...
    _ => {
        if baz {
            return;
        } else {
            a
        }
    }
};
```

I've try to express this in the test code bellow you commented, but yes, as you pointed, that case is not well explaining this and linted as unnecessary.

Thinking about this again, it's tempting to take the method you suggested + extra filtering for tail expression in block. I'll gonna look into this more once I get home

Reply:
For what it's worth, I still think the example above should still be linted and fixed with below 🙂

```rust
let foo = match bar {
    ...
    _ => {
        if baz {
            return;
        }
        a
    }
};
```

Reply:
Haha, I was dumb. You're right 👍

Reply:
Haha, well I'm dumber for writing this diagnostic and not thinking about let statements at all in the first place 😄

Reply:
Not at all. I think that writing something not exists is the hardest thing 😄 
BTW, after applying your suggestions, I'm gonna think about nicer way to fix the triple if-else case you've mentioned. (I'm still at work 😢 )
Thank you for your kind and smart advices 👍